src/librbd/ManagedLock.cc

```diff
@@ -561,7 +561,7 @@ void ManagedLock<I>::send_reacquire_lock() {
   }
 
   m_new_cookie = encode_lock_cookie(watch_handle);
-  if (m_cookie == m_new_cookie) {
+  if (m_cookie == m_new_cookie && m_cookie_valid) {
     ldout(m_cct, 10) << "skipping reacquire since cookie still valid"
                      << dendl;
     complete_active_action(STATE_LOCKED, 0);
```

Comment:
... why not just fire ```post_reacquire_lock_handler``` here instead? Just create a small helper ```handle_no_op_reacquire_lock``` that is invoked by the handler callback (i.e. by ```ExclusiveLock```) that finalizes the state transition to ```STATE_LOCKED```. That way we can still avoid the unnecessary OSD trip.

Reply:
it seems `complete_active_action(STATE_LOCKED, 0);` already avoid the unnecessary OSD trip.
the problem here is we need a OSD trip to reacquire lock,  not avoiding.

so after rewatch, set cookie to be invalid,  then enforce the process to reacquire lock.

Reply:
You actually don't need the OSD trip since the cookie didn't change and the lock still exists. The only thing you need is to invoke the ```post_reacquire_lock_handler``` since that is where ```ExclusiveLock``` re-sends the lock acquired notification (which will reset the owner)

Reply:
but there is a case if the rewatch moment  has exceeded `osd_client_watch_timeout = 30`, the lock will be possibly stealed  by other client. so lock hold may be outdated. then going through the whole procedure of reacquire lock is necessary, not just `post_reacquire_lock_handler` . so the OSD trip should be keeped.

Reply:
@shun-s The client would have been blacklisted by the new lock owner in your case.